### ISSUE TITLE
No way to specify timestamp format reliably (is always milliseconds since 1970)

### ISSUE THREAD

**Issue by zmorris** [agrosner/DBFlow#1521]
DBFlow Version: 3.1.1

Bug or Feature Request: Bug

Description: We have a typical Android application that stores datetimes for created_at/updated_at/deleted_at in TEXT fields in a SQLite database.  Somehow these datetimes are sometimes stored as milliseconds since 1970 like `1516924236000` instead of a datetime like `2018-01-25 23:50:35` and we can't figure out why.

**UPDATE: the problem is not DBFlow, see my answer below: https://github.com/Raizlabs/DBFlow/issues/1521#issuecomment-360652125**

I've tried using my own TypeConverter (note that millisecondsToDateString() and parseStringToDate() are our own internal functions and just do what they say):

```
import com.raizlabs.android.dbflow.converter.TypeConverter;
import java.util.Date;

public class MyTypeConverter extends TypeConverter<String, Date> {


    @Override
    public String getDBValue(Date model) {
        return millisecondsToDateString(model.getTime());	// converts milliseconds to Date and returns string like "YYYY-MM-DD hh:mm:ss"
    }

    @Override
    public Date getModelValue(String data) {
        return parseStringToDate(data);	// converts any date/time/timestamp format to Date
    }
}
```

Usage for a column in MyModel::

```
@Column(name = "created_at", typeConverter = MyTypeConverter.class)
@NotNull
Date added;

@Column(name = "updated_at", typeConverter = MyTypeConverter.class)
@NotNull
Date modified;

@Column(name = "deleted_at", typeConverter = MyTypeConverter.class)
Date deleted;
```

I've also tried overriding the global DatabaseHolder like so:

```
package edu.yale.yins.My.database;

import com.raizlabs.android.dbflow.converter.TypeConverter;
import edu.yale.yins.My.util.Utils;
import java.util.Calendar;

@com.raizlabs.android.dbflow.annotation.TypeConverter
public class MyTypeConverterCalendar extends TypeConverter<String, Calendar> {

    @Override
    public String getDBValue(Calendar model) {
        return model == null ? null : millisecondsToDateString(model.getTimeInMillis());
    }

    @Override
    public Calendar getModelValue(String data) {
        if (data != null) {
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(parseStringToDate(data).getTime());
            return calendar;
        } else {
            return null;
        }
    }
}
```

```
package edu.yale.yins.My.database;

import com.raizlabs.android.dbflow.converter.TypeConverter;
import edu.yale.yins.My.util.Utils;
import java.util.Date;

@com.raizlabs.android.dbflow.annotation.TypeConverter
public class MyTypeConverterDate extends TypeConverter<String, Date> {

    @Override
    public String getDBValue(Date model) {
        return model == null ? null : millisecondsToDateString(model.getTime());
    }

    @Override
    public Date getModelValue(String data) {
        return data == null ? null : parseStringToDate(data);
    }
}
```

```
package edu.yale.yins.My.database;

import com.raizlabs.android.dbflow.converter.TypeConverter;
import edu.yale.yins.My.util.Utils;
import java.sql.Date;

@com.raizlabs.android.dbflow.annotation.TypeConverter
public class MyTypeConverterSqlDate extends TypeConverter<String, Date> {

    @Override
    public String getDBValue(Date model) {
        return model == null ? null : millisecondsToDateString(model.getTime());
    }

    @Override
    public Date getModelValue(String data) {
        return data == null ? null : new java.sql.Date(parseStringToDate(data).getTime());
    }
}
```

```
import com.raizlabs.android.dbflow.converter.BooleanConverter;

public final class MyDatabaseHolder extends DatabaseHolder {
    public MyDatabaseHolder() {
        typeConverters.put(Boolean.class, new BooleanConverter());
        typeConverters.put(java.util.Calendar.class, new MyTypeConverterCalendar());
        typeConverters.put(java.util.Date.class, new MyTypeConverterDate());
        typeConverters.put(java.sql.Date.class, new MyTypeConverterSqlDate());
        new MyDatabaseMy$Database(this);
    }
}
```

```
FlowManager.init(new FlowConfig.Builder(this).build());
FlowManager.initModule(MyDatabaseHolder.class);
```

I've stepped through this in the debugger and verified that `MyDatabaseHolder` is installed.

I've also verified that breakpoints in `MyTypeConverterCalendar`, `MyTypeConverterDate` and `MyTypeConverterSqlDate` are reached as rows are written.

I've also verified that breakpoints in the default DBFlow `CalendarConverter`, `DateConverter` and `SqlDateConverter` are **NOT** reached as rows are written.

So whether specifying TypeConverters for columns individually, or using TypeConverters at a global level, **I still am not able to prevent timestamps getting saved as milliseconds.**

My gut feeling is that when a column is missing or not specified somehow, DBFlow falls back to storing milliseconds.  It's possible that it's also falling back to `GeneratedDatabaseHolder` but I don't see how, because the built-in `CalendarConverter`, `DateConverter` and `SqlDateConverter` are never called.  I've traced the code all the way from `MyModel.save()`, through its adapter, through its `ModelSaver`, and it's saving a datetime string the whole way through.

My question is: does someone know where the final timestamp writing happens in DBFlow so that we can override it?

My other question is: could this be a problem with DBFlow 3.x that has since been fixed?

It's going to be difficult for us to upgrade if this is a 3.x issue but we'll do it if we have to.  Our use case is that we are communicating with a MySQL database online and would like to do everything in a standard SQL datetime format.  Currently we are unable to sort by created_at/updated_at/deleted_at because our timestamps are not being consistently written in the correct format.

So far I've lost 3 days to this learning the intricacies of `FlowManager.init()` and all of the ORM internals with nothing to show for it.  Thanks in advance for any help you can provide.

**Comment by zmorris** [comment]
Apparently, I just had to ask the world and the answer would come.

**The problem was due to a trigger:**

```
CREATE TRIGGER `MyTrigger` AFTER INSERT ON `my_table` BEGIN Update 'my_table' set 'created_at' = (SELECT (strftime('%s', 'now') * 1000)) where id = new.id; END
```

This was written with good intentions by another developer to accomplish something similar to [MySQL timestamp initialization](https://dev.mysql.com/doc/refman/5.7/en/timestamp-initialization.html).

It sounds like this is possible in SQLite https://stackoverflow.com/a/200329/539149 but only in sqlite3 3.1+, although it looks like that's well supported across most versions of Android today https://stackoverflow.com/questions/2421189/version-of-sqlite-used-in-android

```
CREATE TABLE test (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  t TIMESTAMP
  DEFAULT CURRENT_TIMESTAMP
);
```

There seems to be a discrepancy here though in how DBFlow handles default timestamps:

https://github.com/Raizlabs/DBFlow/issues/848

There are arguments for/against settings timestamps in code vs in the database.

I'm going to investigate further, if I find anything more I'll post it here.  Hope this helps someone.
